<a href="https://colab.research.google.com/github/Ben-Ogega/PythonProjects/blob/master/Diabetes_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# This machine learning project is for predicting if someone has diabetes or is likey to be diabetic

### Importing libraries and dependencies

In [246]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

### Data collection and Analysis
### using PIMA diabetes dataset

In [247]:
#Loading the

diabetes_data = pd.read_csv('/content/diabetes.csv')

In [248]:
# Reading the first 5 rows
diabetes_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [249]:
# Check number of rows and Columns in this dataset

In [250]:
diabetes_tuple = diabetes_data.shape

In [251]:
for i in diabetes_tuple:
  print(f'There are {diabetes_tuple[0]} rows and {diabetes_tuple[1]} columns' )
  break

There are 768 rows and 9 columns


In [252]:
# Getting the statistical representation of the data

diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [253]:
# count how many cases are diabetic and how many are not

In [254]:
diabetes_data['Outcome'].value_counts()

0    500
1    268
Name: Outcome, dtype: int64

##### The description above tells us that there are 268 people who are diabetitc while 500 are not.

##### 0 ----> Non diabetic
##### 1 ----> Diabetic

In [255]:
diabetes_data.groupby('Outcome').mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.298000  109.980000      68.184000      19.664000   68.792000   
1           4.865672  141.257463      70.824627      22.164179  100.335821   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.304200                  0.429734  31.190000  
1        35.142537                  0.550500  37.067164

In [256]:
# Separating data

In [257]:
X = diabetes_data.drop('Outcome', axis=1)
X

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  
0                       0.627   50  
1                       0.351   31  
2                       0.672   32  
3                       0.167   21  
4                       2.288   33  
..                        ...  ...  
763                     0.171   63  
764                     0.340   27  
765                     0.245   30  
766                     0.349   47  
767                     0.315   23  

[768 rows x 8 columns]

In [258]:
# Get the label and store it in a variable y

y = diabetes_data['Outcome']
y

0      1
1      0
2      1
3      0
4      1
      ..
763    0
764    0
765    0
766    1
767    0
Name: Outcome, Length: 768, dtype: int64

Data Standardization

In [259]:
scaler = StandardScaler()
#scaler

In [260]:
scaler.fit(X)

StandardScaler()

In [261]:
standardized_data = scaler.fit_transform(X)

In [262]:
standardized_data

array([[ 0.63994726,  0.84832379,  0.14964075, ...,  0.20401277,
         0.46849198,  1.4259954 ],
       [-0.84488505, -1.12339636, -0.16054575, ..., -0.68442195,
        -0.36506078, -0.19067191],
       [ 1.23388019,  1.94372388, -0.26394125, ..., -1.10325546,
         0.60439732, -0.10558415],
       ...,
       [ 0.3429808 ,  0.00330087,  0.14964075, ..., -0.73518964,
        -0.68519336, -0.27575966],
       [-0.84488505,  0.1597866 , -0.47073225, ..., -0.24020459,
        -0.37110101,  1.17073215],
       [-0.84488505, -0.8730192 ,  0.04624525, ..., -0.20212881,
        -0.47378505, -0.87137393]])

In [263]:
X = standardized_data
y = diabetes_data['Outcome']

Train Test Split

In [264]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [265]:
print(X.shape, X_train.shape, X_test.shape)

(768, 8) (614, 8) (154, 8)


Training the Model

In [266]:
classifier = svm.SVC(kernel='linear')

Training the Support Vector Machine

In [267]:
classifier.fit(X_train, y_train)

SVC(kernel='linear')

Evaluating our Model

Accuracy score


In [268]:
# Accuracy score of the training data

In [269]:
X_train_prediction = classifier.predict(X_train)

In [270]:
training_data_accuracy = accuracy_score(X_train_prediction, y_train)

In [271]:
print(f'The accuracy of the model ussing the training data is {training_data_accuracy}')

The accuracy of the model ussing the training data is 0.7915309446254072


Evaluating the model to predict the test data (data the model has never seen)

In [272]:
# Accuracy of the model on the test data

In [273]:
X_test_prediction = classifier.predict(X_test)

In [274]:
testing_data_accuracy = accuracy_score(X_test_prediction, y_test)

In [275]:
print(f'The accuracy of the model using the testing data is {testing_data_accuracy}')

The accuracy of the model using the testing data is 0.7207792207792207


Making a Predictive Model

In [276]:
input_data = (10,125,70,26,115,31.1,0.205,41)

input_data_as_npArray = np.asarray(input_data)



In [277]:
# Reshape the array for one instance
input_data_reshaped = input_data_as_npArray.reshape(1, -1)

# standardize the input data

std_data = scaler.transform(input_data_reshaped)


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  warnings.warn(


In [278]:
#Predict the model

In [279]:
prediction = classifier.predict(std_data)

prediction

array([0])